### ISSUE TITLE
Bundle/integrate/link to ebookmaker

### ISSUE THREAD

**Issue by windymilla** [DistributedProofreaders/guiguts-py#455]
Not sure what will be possible/practical/best. It feels like it ought to be easier to integrate ebookmaker, given that it too is written in python, but that depends on a bunch of stuff I know nothing about related to how ebookmaker is structured. So, there are a few options:

1. Import ebookmaker as a module somehow, and treat it like we do the new pptxt.
2. Install ebookmaker as part of the GG installation, then run it as an external python tool
3. Continue with an ebookmaker.exe on Windows, like GG1, and use method 2 for other platforms, or leave it to the user to sort out.
4. Don't bundle/integrate - PPer will have to use online ebookmaker

Consider whether we should include the W3C Epubcheck tool, like GG1 has.

**Comment by windymilla** [comment]
My initial thought is that option 4 doesn't sound too bad. It has several advantages, including

- online version is always up to date
- reduce size & complication of GG installation
- ebookmaker output does not provide "jump to error" links like home-grown tools
- PPers need to do online ebookmaker check anyway, because the cached output text has to be included when submitting the upload to PG

**Comment by srjfoo** [comment]
I tend to agree that option 4 sounds good. I realize that some PPers like to be able to do everything after they download the PP files offline, but the likelihood that the included version would be stale within a day or three of a GG2 release is pretty high.

**Comment by tangledhelix** [comment]
I also agree about option 4 (and it's what I do myself). Two reasons:

1. Ebookmaker install always seemed to break for me; I gave up trying to maintain it. I'm probably not the only one.
2. Ultimately, our "customer," the PPer, is responsible to produce text and HTML; it's PG who creates the ebook files, and regenerates them periodically. And as we know, there's an effort to produce other formats like PDFs from the same source material. None of this is produced or uploaded by the PP'er directly.

So I think it's appropriate to point to the PG tooling for this. A motivated PP'er can always install Ebookmaker on their own, of course.

**Comment by tangledhelix** [comment]
Revisiting this thought and I'm even more on board for option 4 than before after looking a bit harder. The notion that "ebookmaker is Python" is only true on the surface. Online ebookmaker is doing a number of things, including:

* Generating Kindle files (requires Calibre)
* EpubCheck (Java)
* Nu validator (available in pip as Python but it still seems to require Java)

Could we install only ebookmaker and ignore the other stuff... yes. Is it completely valid to punt and say "Use online ebookmaker" ? Yes... and that's where I'll put my vote, for whatever it counts...

**Comment by srjfoo** [comment]
I'm entirely on board with closing as "not planned" (or the equivalent).

**Comment by srjfoo** [comment]
Revisiting this again.

I'd recommend either closing as not planned, or, since we already have a menu item linking to an external validator, do the same with ebookmaker with the parenthetic (online) in the menu item.

**Comment by windymilla** [comment]
> I'd recommend either closing as not planned, or, since we already have a menu item linking to an external validator, do the same with ebookmaker with the parenthetic (online) in the menu item.

I agree we're not going to include/bundle ebookmaker. The menu item idea is in option 4 in the original issue description above, and I think could be a good idea, perhaps along with one or two other links. It isn't quite the same as the validator link - we use the validator and report the messages to the PPer in a friendly way. For ebookmaker (and other tools) we would just open their browser at the right place and leave them to it.

**Comment by windymilla** [comment]
Completed by #787 providing link to PPWB.

**Comment by jjzsdp** [comment]
You keep wanting to take away things for which we were so eternally grateful when you included them in Guiguts1--the provision of an interim running of ebookmaker is one of them. I don't know about you guys, but when I am PPing and experimenting I can run ebookmaker dozens (+) of times locally. I'm not as excited about having to zip, go to online ebookmaker, upload, wait (and sometimes wait and wait) then having to download to find out I should have made a tiny tweak ... rinse, repeat. Thanks for trying, but I'd certainly appreciate someone trying again to advance this further rather than close the request. (PS Thanks to Nigel's instructions I do have ebookmaker loaded on my system and so far have updated to the current version. Each new instal, thanks to new disks/computer, is fraught, I admit, and that's why having it with guiguts is a god-send.)

**Comment by windymilla** [comment]
> You keep wanting to take away things...

@jjzsdp - I can only assume you don't realise how discouraging it is to read the first sentence of your comment. 

Whether there are good arguments for bundling ebookmaker or not, whether the GG1 method is better or worse, whether you have good ideas on how to do what you want or not - those are all valid points for discussion. We have discussed things like View Options for Bookloupe (which I have implemented); we have discussed making all options available via menus rather than relying on shortcuts (which I have implemented); we have discussed avoiding the need for holding down the Ctrl or Shift key while clicking (which I have implemented)

"You keep wanting to take away things..." is not only upsetting to read, it's factually untrue, unless you have some evidence to the contrary.

**Comment by windymilla** [comment]
> I do have ebookmaker loaded on my system and so far have updated to the current version. Each new instal, thanks to new disks/computer, is fraught, I admit,

Since GG2 users will already have Python installed, if it's not easy to install ebookmaker, that is really a problem that might more profitably be raised with the ebookmaker maintainer (who is doing an excellent job already in my opinion). I have made GG2 very easy to install once you have python, via a single command. There may be good technical reasons why you find ebookmaker harder to install - I don't know, because I don't work on ebookmaker.

Since you don't know how hard it was to get ebookmaker bundled with GG1 on Windows only (I needed lots of expert help, and was unable to do it by myself) nor how much work it is to adapt the process when certain types of change are made in ebookmaker, nor how much additional work it takes to create a GG1 release because of ebookmaker, I think you could have chosen your words more carefully, rather than just stating that I "keep wanting to take away things", with an implied "without good reason". I don't **_want_** to take away things, but I am trying to balance a lot of different requirements, some of which I have mentioned previously, and one of which is my life outside of GG2 development.

**Comment by jjzsdp** [comment]
I do understand how jolly hard you are working on trying to meet everyone's needs and I can't express how much I appreciate everything single minute you have invested in the task. I regret my words were not interpreted as they were intended. I wrote elsewhere and will repeat here for the record, it's your fault, Nigel, you did such an incredibly brilliant job of dragging guiguts1 up to the needs of its users, that if guiguts2 doesn't have similar features and measure up to what you've given us, we will lament the loss. It's not that I want it to be exactly like guiguts1, but rather I would hope it would be superior, not missing some cogs that I've been relying on. Those who want to, and/or are able to, can follow your directions to get ebookmaker working locally, at least on Windows, so your suggestion that it might be possible to link it up to GG2 so you can run it from a menu entry sounds great, if/when you are able to spin the magic for that to happen. If not, so be it. Thank you, and everyone contributing.

**Comment by windymilla** [comment]
Thanks for the encouragement @jjzsdp. We are aware that it's frustrating sometimes for users when developers either choose to, or need to, do things differently to the ways they have been done previously. With patience and continued communication, I hope that we can address most of what users need, and appreciate your understanding when we can't. I'll add a "future feature" issue to look at what can be done to make GG2 able to utilize ebookmaker if it's installed on the user's computer.
